dapp_runner/runner/runner.py

```diff
@@ -230,18 +230,20 @@ def dapp_state(self) -> Dict[str, Dict[int, ServiceState]]:
         Clusters and their Service instances comprising the dapp.
         """
 
-        return {
+        cluster_states = {
             cluster_id: {
                 instance_index: instance.state
                 for instance_index, instance in enumerate(cluster.instances)
             }
-            if cluster
-            else {}
-            for cluster_id, cluster in [
-                (cluster_id, self.clusters.get(cluster_id)) for cluster_id in self.dapp.nodes.keys()
-            ]
+            for cluster_id, cluster in self.clusters.items()
         }
 
+        missing_nodes = set(self.dapp.nodes.keys()).difference(cluster_states.keys())
```

Comment:
This could be simplified.

```suggestion
        missing_nodes = set(self.dapp.nodes) - set(cluster_states)
```